### ISSUE TITLE
NullPointerException for several quantified instance( incremental mode)

### ISSUE THREAD

**Issue by rainoftime** [ultimate-pa/smtinterpol#50]
Hi, for the following formula
~~~~
(set-logic NRA)
(declare-fun v0 () Bool)
(declare-fun v1 () Bool)
(declare-fun v2 () Bool)
(declare-fun v3 () Bool)
(declare-fun v4 () Bool)
(declare-fun v5 () Bool)
(declare-fun v6 () Bool)
(declare-fun v7 () Bool)
(declare-fun r0 () Real)
(declare-fun r1 () Real)
(declare-fun r2 () Real)
(assert (not v3))
(check-sat)
(assert (or (>= r1 r2 r0) (>= r1 r2 r0) v5 v2 v2))
(declare-fun v10 () Bool)
(declare-fun v11 () Bool)
(assert (not (exists ((q3 Bool) (q4 Bool) (q5 Real) (q6 Bool)) (=> (>= q5 r1) (xor q4 q3 q3 v10 v11 q3 (or v4 (xor v7 v5 v7) v3 v5 v3 (not v3) v7 v4 v10))))))
(push 1)
(push 1)
(check-sat)
(pop 1)
(push 1)
(check-sat)
~~~~

smtinterpol (commit 3f24148180a10d8) throws a NullPointerException
~~~~
Unexpected Exception: java.lang.NullPointerException
(error "java.lang.NullPointerException")
~~~~

**Comment by rainoftime** [comment]
Here I append test cases from other similar issues 

NIA from https://github.com/ultimate-pa/smtinterpol/issues/53
~~~~
(set-logic NIA)
(declare-fun v0 () Bool)
(declare-fun v1 () Bool)
(declare-fun v2 () Bool)
(declare-fun v5 () Bool)
(declare-fun v11 () Bool)
(declare-fun v12 () Bool)
(declare-fun v17 () Bool)
(declare-fun v18 () Bool)
(declare-fun i2 () Int)
(declare-fun i3 () Int)
(declare-fun i4 () Int)
(push 1)
(assert (exists ((q0 Int) (q1 Bool) (q2 Bool) (q3 Int)) (=> (or v17 q1 q2 q2 v0) (< q3 q0))))
(declare-fun v20 () Bool)
(assert (not (exists ((q4 Bool) (q5 Bool)) (= (< 35 i4) v0 v12))))
(assert (or (forall ((q4 Bool) (q5 Bool)) v18) (exists ((q4 Bool) (q5 Bool)) (not (or v5 q5 q5 q4 q5)))))
(declare-fun v21 () Bool)
(assert (forall ((q6 Bool)) (not (xor q6 v0 q6 v11 q6 q6 (or v21 (> i4 29) v21 v12) (< 35 i4) q6))))
(push 1)
(assert (not (exists ((q7 Int)) (> i4 29))))
(pop 1)
(check-sat)
~~~~

**Comment by rainoftime** [comment]
LIA formula from https://github.com/ultimate-pa/smtinterpol/issues/55
~~~~
(set-logic LIA) 
(declare-fun v0 () Bool)
(declare-fun v1 () Bool)
(assert (forall ((q0 Bool) (q1 Bool)) (not (xor q1 v1 v0 q0 q1 (> 100 13) v0 q1))))
(push 1)
(assert (=> v1 v1))
(pop 1)
(check-sat)
~~~~

**Comment by rainoftime** [comment]
UF formula from https://github.com/ultimate-pa/smtinterpol/issues/57
~~~~~
(set-logic UF)
(declare-fun v0 () Bool)
(declare-fun v3 () Bool)
(declare-fun v16 () Bool)
(assert (not (exists ((q4 Bool) (q5 Bool) (q6 Bool)) (not (xor q5 v16 q4 v0 (xor v16 v3) q5)))))
(push 1)
(check-sat)
(pop 1)
(check-sat)
~~~~

**Comment by jhoenicke** [comment]
All problems seem related.  It looks like the quantifier theory propagates a literal that was already removed by a previous pop().

@tanjaschindler my guess is that the InstClause caches these literals.  One should probably remove all InstClauses on a pop, even if the quantified clause is not removed.

**Comment by rainoftime** [comment]
Here is the test case from https://github.com/ultimate-pa/smtinterpol/issues/52 (the type of exception is different)
~~~~
(set-logic NIA)
(declare-fun v0 () Bool)
(declare-fun v1 () Bool)
(declare-fun v2 () Bool)
(declare-fun v4 () Bool)
(declare-fun v5 () Bool)
(declare-fun v7 () Bool)
(declare-fun i0 () Int)
(declare-fun i1 () Int)
(declare-fun i5 () Int)
(declare-fun i7 () Int)
(assert v1)
(declare-fun v8 () Bool)
(assert (< 59 61))
(check-sat)
(declare-fun v10 () Bool)
(assert (=> v10 (< 52 i1)))
(assert (not (exists ((q0 Bool) (q1 Int) (q2 Int)) (not (>= 61 q2)))))
(push 1)
(push 1)
(assert (exists ((q7 Int)) (and v10 v4 v1 (xor (distinct i7 i5) v2 (=> v10 (< 52 i1)) v10 v4 v8 v7 v10 v0 v4 v5))))
(assert (forall ((q9 Bool) (q10 Int) (q11 Int)) (not (not q9))))
(pop 1)
(check-sat)
(push 1)
(check-sat)
~~~~

**Comment by rainoftime** [comment]
Test case from https://github.com/ultimate-pa/smtinterpol/issues/58
~~~~
(set-logic LIA) 
(declare-fun v4 () Bool)
(declare-fun v7 () Bool)
(declare-fun v10 () Bool)
(assert (not (exists ((q0 Bool) (q1 Int) (q2 Bool) (q3 Int)) (not (>= 6 q3)))))
(push 1)
(assert (xor v7 v10 (= v4 v7 ) v4 (not v10)))
(pop 1)
(check-sat)
~~~~

**Comment by tanjaschindler** [comment]
Except for the UF file, this should be solved by d768667

**Comment by tanjaschindler** [comment]
For the UF file, this should be solved by 5171d77
